### PR TITLE
Add garbage collection to jobs

### PR DESCRIPTION
**Pull request by dadgar**
This PR adds garbage collection to jobs and marks batch jobs as eligible for garbage collection

### CODE REVIEW

**Comment by diptanu** [comment]
Some thoughts about the implementation -

a. I am wondering why we have exposed the gc duration configurable for an end user of Nomad. The only guarantee Nomad needs to provide to an end user whose job is running is a. highly available scheduling b. jobs getting scheduled when resources are available. GCing jobs out of Nomad when they are no longer going to be referenced by the system doesn't feel like something an end user needs to configure via the job defn.

b. We could just GC any job which won't be referenced again in the system. The only jobs that feels GC-able to me are batch jobs. Service jobs which has count > 1 and num of Tasks > 1 should never be GC-able until the moment the user stops the job. Setting count = 0 or num Tasks = 0 is almost same as the job getting stopped. System jobs, the way they are currently implemented shouldn't be GC-ed as well.

c. Periodic Jobs are interesting - If a periodic job won't be invoked again to submit a child jobs because of the configuration of it's spec, it's a candidate for being GC-ed, otherwise there shouldn't be any need for GC-ing it. 

I think (b) and (c) just explains that when a job can be GC-ed from the system could be inferred by the state of the current allocations and whether any future evaluations would be created for that job without needing any additional configuration from the end user whose job is running on Nomad.

**Comment by dadgar** [comment]
@diptanu: Updated to address comments.

File: nomad/core_sched.go [hunk 1]
```diff
@@ -33,11 +33,90 @@ func (s *CoreScheduler) Process(eval *structs.Evaluation) error {
...
+	}
+
+	// Get the time table to calculate GC cutoffs.
+	tt := c.srv.fsm.TimeTable()
+	cutoff := time.Now().UTC().Add(-1 * c.srv.config.JobGCThreshold)
+	oldThreshold := tt.NearestIndex(cutoff)
+	c.srv.logger.Printf("[DEBUG] sched.core: job GC: scanning before index %d (%v)",
+		oldThreshold, c.srv.config.JobGCThreshold)
+
+	// Collect the allocations and evaluations to GC
```
Comment (diptanu) [inline nomad/core_sched.go]: allocations, evaluations and jobs

File: nomad/core_sched.go [hunk 2]
```diff
@@ -100,10 +156,52 @@ OUTER:
...
+
+// gcEval returns whether the eval should be garbage collected given a raft
+// threshold index. The eval disqualifies for garbage collection if it or its
+// allocs are not older than the threshold. If the eval should be garbage
+// collected, the associated alloc ids that should also be removed are also
+// returned
+func (c *CoreScheduler) gcEval(eval *structs.Evaluation, thresholdIndex uint64) (
+	bool, []string, error) {
+	// Ignore non-terminal and new evaluations
+	if !eval.TerminalStatus() || eval.ModifyIndex > thresholdIndex {
```
Comment (diptanu) [inline nomad/core_sched.go]: Put this logic in eval? 

```
func (e *Evaluation) IsGCable(thresholdIdx int) {
}
```

File: nomad/structs/structs.go [hunk 3]
```diff
@@ -1474,14 +1490,21 @@ type Allocation struct {
...
-// will no longer transition. This is not based on the current client status.
+// TerminalStatus returns if the desired or actual status is terminal and
+// will no longer transition.
 func (a *Allocation) TerminalStatus() bool {
 	switch a.DesiredStatus {
 	case AllocDesiredStatusStop, AllocDesiredStatusEvict, AllocDesiredStatusFailed:
 		return true
 	default:
-		return false
+		// If all tasks are dead, the alloc is terminal.
```
Comment (diptanu) [inline nomad/structs/structs.go]: Do we need this logic here? I think ideally the client is going to do this and set the Alloc Client status.

**Comment by diptanu** [comment]
LGTM overall.

File: nomad/core_sched.go (hunk 2 above)
Comment (dadgar) [inline nomad/core_sched.go]: Its pretty specific to what we are doing here and wouldn't really be useful anywhere else.

File: nomad/structs/structs.go (hunk 3 above)
Comment (dadgar) [inline nomad/structs/structs.go]: Good call

**Pull request merged**